src/wasm2asm.h

```diff
@@ -228,17 +230,34 @@ class Wasm2AsmBuilder {
 Ref Wasm2AsmBuilder::processWasm(Module* wasm) {
   PassRunner runner(wasm);
   runner.add<AutoDrop>();
-  runner.add("remove-non-js-ops"); // must be before i64-to-i32
+  // First up remove as many non-JS operations we can, like 64-bit integer
+  // multiplication or division. This'll make it possible actually perform
+  // i64-to-i32 lowering as otherwise that pass can't handle things like
+  // rotation/division.
+  runner.add("remove-non-js-ops");
   // Currently the i64-to-32 lowering pass requires that `flatten` be run before
   // it produce correct code. For some more details about this see #1480
   runner.add("flatten");
   runner.add("i64-to-i32-lowering");
```

Comment:
i don't quite understand the interaction between these two passes. it seems like you're saying we need to
 * remove non-js ops, like an i64 div
 * lower i64 ops to i32 ops
 * remove non-js ops again

what is an example of something that needs the last line?

Reply:
Sure yeah, the example here is the `i64.ctz` instruction. The first remove-non-js-ops pass doesn't remove it (as it doesn't have anything to lower to currently). The i64-to-i32 pass then lowers that to adding the result of `i32.ctz` on each half. The wasm2asm pass, however, doesn't know how to implement `i32.ctz` so it aborts.

In other words, the first pass runs to remove operations that may use i64, like the lowering of `i64.div_s`. The second pass then removes i64 as a type everywhere. Finally if the i64 removal injected any instructions that we still don't know what to do with they're cleaned up by remove-nonjs-ops again

Reply:
Ah yeah it's basically arbitrary right now. We could add intrinsics for i64.ctz but I figured it was easiest to rely on the existing pass to split that and probably slightly more efficient codegen which calls i32.ctz twice instead of i64.ctz, and then I figured it was best to leave it all in one pass instead of splitting it off

Reply:
I see, thanks.

This arbitrariness worries me regarding maintaining and improving the code. Let's think some more on this. My current thinking is that we could

 * Lower all i64 instructions in one pass. This needs the temp var logic and all that.
 * Lower all non-i64 instructions that are non-JS (like `f32.copysign`) in another. This doesn't need temp vars.

And both might need an intrinsics module that implements things we replace, so we might want some shared helper code for them.

What do you think?

Reply:
Alternatively, the two might just be one pass for all the non-JS stuff. I don't see a reason to only lower i64s and not all non-JS operations, myself.

Reply:
I suppose we could organize it like that? I don't really see though how it'd be cleaner or maintainable per se than what's currently there. It seems like it'd be taking the current intrinsic injection pass and basically splitting it across two passes because it needs to happen in two locations. 

Is there a downside to running the pass twice?

Reply:
There is a minor speed downside to running it twice, but I think we can live with it. My concern is that arbitrary design harms maintainability in my experience. For example, if some time from now I want to add another method to be lowered (say, after wasm adds GC or multi-value) then it would be good to have a clear, proper place for it. If there are two equally valid passes, I'd need to understand both passes before making that decision.

Another example is if a bug occurs, running the pass twice could make things somewhat more complicated to debug, as the bug could occur in either one. And future additions to either pass might need to take into account that one of them runs twice, etc.

Reply:
I think though the only reason that this could be construed as arbitrary is in the sense that the i64.ctz intrinsic isn't lowered and instead it only lowers the i32.ctz intrinsic. If you'd prefer I could add intrinsic calls for both of those and we wouldn't have to run it twice.

Otherwise I don't think the purpose of each pass is really that arbitrary. The i64-to-i32 pass successfully eliminates all 64-bit integers, it just panics if it hits things like division. The remove-non-js-ops successfully injects a bunch of intrinsics, it just doesn't handle everything it possibly can.

Aren't there already required orderings as well? For example if flatten isn't run before i64-to-i32 then broken code is produced?